Draft extension for SPZ compression support for Gaussian splats (#81)
CesiumGS/glTF: extensions/2.0/Khronos/KHR_spz_gaussian_splats_compression/README.md

| Rotation | _ROTATION | VEC4 | float | yes | Rotation is a quaternion. |
| Scale | _SCALE | VEC3 | float | yes | |
| Spherical Harmonics degree 1 | _SH_DEGREE_1_COEF_n (n = 0 to 2) | VEC3 | float | no (yes if degree 2 or 3 are used) | |
| Spherical Harmonics degree 2 | _SH_DEGREE_2_COEF_n (n = 0 to 4) | VEC3 | float | no (yes if degree 3 is used) | |
| Spherical Harmonics degree 3 | _SH_DEGREE_3_COEF_n (n = 0 to 6) | VEC3 | float | no | |

Each increasing degree of spherical harmonics requires more coeffecients. At the 1st degree, 3 sets of coeffcients are required, increasing to 5 sets for the 2nd degree, and increasing to 7 sets at the 3rd degree. With all 3 degrees, this results in 45 spherical harmonic coefficients stored in the `_SH_DEGREE_ℓ_COEF_n` attributes.

**TODO:** SPZ to attributes diagram

### Accessors

Required `accessors` for `POSITION`, `COLOR_0`, `_ROTATION`, and `_SCALE`:

```json
  "accessors": [{
      "componentType": 5126,
      "count": 590392,
      "type": "VEC3",
      "max": [
        1,
        1,
        1,
      ],
      "min": [
        -1,
        -1,
        -1,

>       ]

Reviewer: This is required by the glTF spec for all `POSITION` accessors.

Reviewer: > A good ask for the working group is if we can relax this requirement when our extension is in use.

I changed my mind about this. I think there's still potential for these bounds to be very useful in some engines. We should still require these bounds to be properly calculated.

    }, {
      "componentType": 5121,
      "count": 590392,
      "type": "VEC4",

>       "normalized": true

Reviewer: Colors represented by bytes are required to be normalized per spec.

    }, {
      "componentType": 5126,
      "count": 590392,

>       "type": "VEC3"

Reviewer 2: If this example is showing rotation then scale, it should be `VEC4` for rotation.

```suggestion
      "type": "VEC4"

```

Reviewer: Actually, these two accessors are swapped in the statement on line 124. I want to make sure that the order corresponds to the example given on line 67.

    }, {
      "componentType": 5126,
      "count": 590392,

>       "type": "VEC4"

Reviewer 2: If this example is showing rotation then scale, it should be `VEC3` for scale.

```suggestion
      "type": "VEC3"

```

    }],
```

Spherical harmonics `accessors` all follow the pattern:

```json
  "accessors": [{
    "componentType": 5126,
    "count": 590392,

>     "type": "SCALAR"

Reviewer 2: Spherical harmonic accessors are now `VEC3`

```suggestion
    "type": "VEC3"

```

  }]
```

At minimum accessors will be defined for `POSITION`, `COLOR_0`, `_ROTATION`, and `_SCALE`. Each must have `componentType`, `count`, and `type` defined.

Accessor `type` is defined for the resulting type after decompression and dequantization has occurred.

> The accessor `count` must match `numPoints` in `KHR_spz_gaussian_splats_compression`.

Reviewer 2: `numPoints` was removed. So maybe this should say:

```suggestion
The accessor `count` must match the number of points in the compressed SPZ data.
```

## Conformance

The recommended process for handling SPZ compression is as follows:

- If the loader does not support `KHR_spz_gaussian_splats_compression`, it must fail.
- If the loader does support `KHR_spz_gaussian_splats_compression` then:

  - The loader must process `KHR_spz_gaussian_splats_compression` data first. The loader must get the data from `KHR_spz_gaussian_splats_compression`'s `bufferView` property.
  - The loader then must process `attributes` of the `primitive`. When processing the loader must ignore any `bufferView` and `byteOffset` in the `accessor` and instead use values derived from the decompressed data streams. This data can be used to populate the `accessors` or render directly.

>   - Any attributes not listed in `KHR_spz_gaussian_splats_compression`'s `attributes` but are listed in the `primitive` must then be processed by the loader.

Reviewer 2: The extension object no longer has `attributes` so maybe this note can be removed?

Reviewer: Yes. Instead, we probably need to mention here how the mapping should occur. Adding this now.

The extension must be listed in `extensionsUsed`

```json
  "extensionsUsed" : [
    "KHR_spz_gaussian_splats_compression"
  ]
```

It must also be listed in `extensionsRequired`

```json
  "extensionsRequired" : [
    "KHR_spz_gaussian_splats_compression"
  ]
```

Note that while `KHR_spz_gaussian_splats_compression` is required.

### Geometry Type

The `mode` of the `primitive` must be `POINTS`.

### Schema Example

Example SPZ extension shown below. This extension only affects any `primitive` nodes containting Gaussian splat data.

```json
  "meshes": [{
      "primitives": [{
          "attributes": {
            "_ROTATION": 3,
            "COLOR_0": 1,
            "_SCALE": 2,
            "POSITION": 0,
            "_SH_DEGREE_1_COEF_0": 4,
            "_SH_DEGREE_1_COEF_1": 5,
            "_SH_DEGREE_1_COEF_2": 6,
          },
          "material": 0,
          "mode": 0,
          "extensions": {
            "KHR_spz_gaussian_splats_compression": {
              "bufferView": 0,
              "numPoints": 590392,
              "shDegree": 0,
              "fractionalBits": 12,

>               "flags": 0,

Reviewer 2: I'm wondering which of these are actually needed:

* `numPoints` - seems superfluous when there's also accessor `count`
* `shDegree` - could this be determined by the attributes that are present?
* `fractionalBits` - how important is this / how is it used at runtime?

Author: Right now it's really just informational. A way to know what is in the SPZ header without decompressing anything.  `fractionalBits` is used for position quantization. While `numPoints` is a bit redundant with `count` and `shDegree` could be determined by attributes, it's just reflecting the SPZ header info.

To me it would be ideal to have these values in the glTF and not write a header into the compressed stream in the future.

Reviewer 2: > To me it would be ideal to have these values in the glTF and not write a header into the compressed stream in the future.

In the meantime it might be best to remove the header details from the extension. Thinking about the draco extension, it also has a header in the compressed stream but doesn't expose any of those details in the glTF extension.

Reviewer 2: So the spz extension may just consist of a `bufferView` property

Reviewer: Honestly, removing them makes a ton of sense to me. We don't need them for decompressing the data, so they probably shouldn't live in the glTF metadata.

              "attributes": {
                "_ROTATION": 3,
                "COLOR_0": 1,
                "_SCALE": 2,
                "POSITION": 0,
                "_SH_DEGREE_1_COEF_0": 4,
                "_SH_DEGREE_1_COEF_1": 5,
                "_SH_DEGREE_1_COEF_2": 6,
              }
            }
          }
        }]
    }],
  "buffers": [{

>       "uri": "0.spz",

Reviewer 2: .bin seems more appropriate in a glTF context

```suggestion
      "uri": "0.bin",

```

      "byteLength": 9753142
    }],
  "bufferViews": [{
      "buffer": 0,
      "byteLength": 9753142
    }],
```

### Extension Properties

#### bufferView

This property points to the bufferView containing the Gaussian splat data compressed with SPZ.

#### attributes

This contains the attributes that will map into the compressed SPZ data indicated by `bufferView`. At minimum it will contain `POSITION`, `COLOR_0`, `_ROTATION`, and `_SCALE`. `_SH_DEGREE_ℓ_COEF_n` attributes hold the spherical harmonics data and are not required. If higher degrees are used then lower degrees are required implicitly.

| Splat Data | glTF Attribute | Accessor Type | Component Type | Required |
| --- | --- | --- | --- | --- |
| Position | POSITION | VEC3 | float | yes |

> | Color (Spherical Harmonic degree 0 (Diffuse)) | COLOR_0 | VEC4 | unsigned byte or float normalized | yes |

Reviewer 2: Would be good to mention alpha somewhere

```suggestion
| Color (Spherical Harmonic degree 0 (Diffuse) and alpha) | COLOR_0 | VEC4 | unsigned byte or float normalized | yes |

```

> | Color (Spherical Harmonic degree 0 (Diffuse)) | COLOR_0 | VEC4 | unsigned byte or float normalized | yes |

Reviewer 2: I think this would be more accurate

```suggestion
| Color (Spherical Harmonic degree 0 (Diffuse)) | COLOR_0 | VEC4 | unsigned byte normalized or float| yes |

```

Reviewer: I'm not sure what is correct here. @keyboardspecialist, can you take a look?

Author: This change is correct according to the glTF 2.0 spec.

> | Rotation | _ROTATION | VEC4 | float | yes |

Reviewer 2: I'm not sure if it's mentioned anywhere that `_ROTATION` is a quaternion

Reviewer: It isn't, but it should be. I've fixed it.

### Accessors

Required `accessors` for `POSITION`, `COLOR_0`, `_ROTATION`, and `_SCALE`:

```json
  "accessors": [{
      "componentType": 5126,
      "count": 590392,
      "type": "VEC3"
    }, {
      "componentType": 5121,
      "count": 590392,
      "type": "VEC3"
    }, {
      "componentType": 5126,
      "count": 590392,
      "type": "VEC3"
    }, {
      "componentType": 5126,
      "count": 590392,
      "type": "VEC4"
    }],
```

Spherical harmonics `accessors` all follow the pattern:

```json
  "accessors": [{
    "componentType": 5126,
    "count": 590392,
    "type": "SCALAR"
  }]
```

At minimum accessors will be defined for `POSITION`, `COLOR_0`, `_ROTATION`, and `_SCALE`. Each must have `componentType`, `count`, and `type` defined.

Accessor `type` is defined for the resulting type after decompression and dequantization has occurred.

The accessor `count` must match `numPoints` in `KHR_spz_gaussian_splats_compression`.

## Conformance

The recommended process for handling SPZ compression is as follows:

- If the loader does not support `KHR_spz_gaussian_splats_compression`, it must fail.

> - If the load does support `KHR_spz_gaussian_splats_compression` then:

Reviewer 2: ```suggestion
- If the loader does support `KHR_spz_gaussian_splats_compression` then:

```
